### PR TITLE
New Rule: AvoidMutableCollectionInParallelStreamForEach #408

### PR DESCRIPTION
**Pull request by jborgers**

### CODE REVIEW

File: src/test/resources/com/jpinpoint/perf/lang/java/ruleset/concurrent/xml/AvoidMutableCollectionInParallelStreamForEach.xml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    }
+    public void bad3() {
+        Map<String, String> upperToUppercase = new HashMap<>();
+
+        names.parallelStream().map(String::toUpperCase) // bad 3
+                .forEach(name -> upperToUppercase.put(name, name));
+    }
+
+    public void good2(List inUppercase) { // don't know how implemented, could be thread-safe
```
Comment (stokpop) [inline src/test/resources/com/jpinpoint/perf/lang/java/ruleset/concurrent/xml/AvoidMutableCollectionInParallelStreamForEach.xml:53]: It is actually bad practice to modify a "foreign" list from a non private method... in practice you can also not trust that a thread safe list is provided: could recommend to make a copy and return a (new) list instead?

**Review by stokpop — APPROVED** [review]
Good to have this rule! 

Seems ok, but I have some questions/discussions about possible factory methods to add and about how to deal with possible modifiable collections in parameters.

File: rulesets/java/jpinpoint-rules.xml [hunk 2]
```diff
@@ -2934,6 +2935,69 @@ public static String good(CompletableFuture<String> complFuture) throws Exceptio
...
+        <properties>
+            <property name="tags" value="jpinpoint-rule,multi-threading,performance,pitfall" type="String" description="classification"/>
+            <property name="xpath">
+                <value><![CDATA[
+(: should match java.lang.Stream#forEach(java.util.function.Consumer) but also others like java.util.stream.IntStream#forEach(java.util.function.IntConsumer) :)
+//MethodCall[pmd-java:matchesSig('java.util.stream.BaseStream#forEach(_)')][.//MethodCall[@MethodName='parallel' or @MethodName='parallelStream']]
+[ArgumentList[
+    .//(LambdaExpression|MethodReference)[
+        (: a constructor of a thread-unsafe collection/map :)
+        .//VariableAccess/@Name = (ancestor::MethodDeclaration//LocalVariableDeclaration|ancestor::ClassBody/FieldDeclaration)/VariableDeclarator[
```
[the remark this answers is not in the document]
Comment (jborgers) [reply rulesets/java/jpinpoint-rules.xml:2957]: Indeed the .of methods return a thread-safe, immutable collection.
The Collectors.toXX as documented don't guarantee anything concerning mutability or thread-safety., yet, they seem to be mutable. 
Does this occur in actual code?
If so, maybe create a fix request.

File: src/test/resources/com/jpinpoint/perf/lang/java/ruleset/concurrent/xml/AvoidMutableCollectionInParallelStreamForEach.xml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    public void good() {
+        // pure functional form
+        List<String> inUppercase = names.parallelStream().map(String::toUpperCase).toList(); // java 12+
+    }
+
+    public void bad2() {
+        String[] inUppercase = new String[names.size()];
+        AtomicInteger i = new AtomicInteger(0);
+        names.parallelStream().map(String::toUpperCase) // bad 2
```
[the remark this answers is not in the document]
Comment (jborgers) [reply src/test/resources/com/jpinpoint/perf/lang/java/ruleset/concurrent/xml/AvoidMutableCollectionInParallelStreamForEach.xml:42]: that would be nice, but bcos of the chaining, pmd always ends up at the first line of chaining.

File: src/test/resources/com/jpinpoint/perf/lang/java/ruleset/concurrent/xml/AvoidMutableCollectionInParallelStreamForEach.xml (hunk 1 above)
Comment (jborgers) [reply src/test/resources/com/jpinpoint/perf/lang/java/ruleset/concurrent/xml/AvoidMutableCollectionInParallelStreamForEach.xml:53]: Right, that would be another rule, it is suspicious. If this occurs more than once in actual code, it makes sense to create such a rule in my view

**Review by stokpop — APPROVED** [review]

**Pull request merged**